src/client.rs

```diff
@@ -912,16 +915,64 @@ impl Client {
         // Imagine if a cookie is set with path=/download/some_identifier. How do we get that
         // cookie without triggering a request for the (large) file? I don't know. Hence: TODO.
         let old_url = self.current_url_().await?;
-        let url = old_url.clone().join(&url)?;
-        let cookie_url = url.clone().join("/please_give_me_your_cookies")?;
-        self.goto(cookie_url.as_str()).await?;
+        let cookie_url = old_url.join(url)?.join("/please_give_me_your_cookies")?;
+        self.with_raw_cookie_client_for(method, url, Some(cookie_url.as_str()), before)
+            .await
+    }
+
+    /// Issue an HTTP request to the given `url` with all the same cookies as the current
+    /// session. If set, the client will first navigate to
+    /// `cookie_url` to retrieve cookies.
+    ///
+    /// Calling this method is equivalent to calling `with_raw_cookie_client_for` with an empty closure.
+    ///
+    /// `raw_client_for` is an alternative that automatically handles cookies.
+    pub async fn raw_cookie_client_for(
```

Comment:
I'm not in love with the name here. I wonder if the way to go rather than having all of these different methods is to do a builder instead. I'm thinking:
```rust
impl Client {
    fn raw_request(&self) -> RawRequestBuilder;
}
impl RawRequestBuilder {
    fn method(&mut self, m: Method) -> &mut Self;
    fn url(&mut self, u: &str) -> &mut Self;
    fn cookie_url(&mut self, u: &str) -> &mut Self;
    fn map_request(&mut self, f: impl FnOnce(http::request::Builder) -> hyper::Request<hyper::Body>) -> &mut Self;
    fn send(self) -> Result<hyper::Response<hyper::Body>, error::CmdError>;
}
```
Given that we're going to do a breaking release anyway, fixing this properly now seems like a good idea. What do you think @stevepryde?

Reply:
Yeah the builder idea is nice 👍 It should also allow for more capability than before (and future expansion) without adding more methods to `Client`.

I also think the original issue has merit as well - hitting a "random" url before the one specified by the caller does seem like surprising behaviour. I get why this is needed in order to get the cookies, but I think this needs to be more explicit. Something like the builder would make it more explicit in the calling code. The cookie fetch behaviour should also be documented as well (as it is in this PR currently).

Note that this could also allow the caller to NOT include cookies if they want (just skip the `cookie_url()`, thus avoiding the extra request) and then they have convenient access to a http client for any other use case where the cookies might be unnecessary or problematic. I'm not sure if this is useful or something you want to support. It's probably out of scope here at least.

Reply:
Sweet, I'll convert this to a builder!

Reply:
There are three cases here:

1. Don't hit a URL but send current cookies
2. Hit a URL and get cookies
3. Don't hit a URL and don't send current cookies

I think the builder should support all three.

Reply:
I agree